packages/playwright/src/runner/watchMode.ts

```diff
@@ -19,25 +19,77 @@ import { createGuid, getPackageManagerExecCommand, ManualPromise } from 'playwri
 import { enquirer } from '../utilsBundle';
 import { separator } from '../reporters/base';
 import { PlaywrightServer } from 'playwright-core/lib/remote/playwrightServer';
-import ListReporter from '../reporters/list';
+import { TestServerDispatcher } from './testServer';
+import { EventEmitter } from 'stream';
+import { type TestServerSocket, TestServerConnection } from '../isomorphic/testServerConnection';
 
-class FSWatcher {
-  private _dirtyTestFiles = new Map<FullProjectInternal, Set<string>>();
-  private _notifyDirtyFiles: (() => void) | undefined;
-  private _watcher: CFSWatcher | undefined;
-  private _timer: NodeJS.Timeout | undefined;
+class InMemoryServerSocket extends EventEmitter implements TestServerSocket {
+  constructor(public readonly send: (data: any) => void, public readonly close: () => void = () => {}) {
+    super();
+  }
+
+  addEventListener(event: string, listener: (e: any) => void) {
+    this.addListener(event, listener);
+  }
+}
+
+export async function runWatchModeLoop(config: FullConfigInternal): Promise<FullResult['status']> {
+  const testServerDispatcher = new TestServerDispatcher({ configDir: config.configDir, resolvedConfigFile: config.config.configFile });
+  const inMemorySocket = new InMemoryServerSocket(
+      async data => {
+        const { id, method, params } = JSON.parse(data);
+        try {
+          const result = await testServerDispatcher.transport.dispatch(method, params);
+          inMemorySocket.emit('message', { data: JSON.stringify({ id, result }) });
+        } catch (e) {
+          inMemorySocket.emit('message', { data: JSON.stringify({ id, error: String(e) }) });
+        }
+      }
+  );
+  testServerDispatcher.transport.sendEvent = (method, params) => {
+    inMemorySocket.emit('message', { data: JSON.stringify({ method, params }) });
+  };
+  const testServerConnection = new TestServerConnection(inMemorySocket);
+  inMemorySocket.emit('open');
+
+  const dirtyTestFiles = new Map<FullProjectInternal, Set<string>>();
+  const onDirtyTestFiles: { resolve?(): void } = {};
+  const failedTestIdCollector = new Set<string>();
+
+  testServerConnection.onTestFilesChanged(({ testFiles }) => onTestFilesChanged(testFiles));
+  testServerConnection.onReport(report => {
+    if (report.method === 'onTestEnd') {
+      const { result: { status }, test: { testId, expectedStatus } } = report.params;
+      if (status !== expectedStatus)
+        failedTestIdCollector.add(testId);
+      else
+        failedTestIdCollector.delete(testId);
+    }
+  });
+
+  const originalStdoutWrite = process.stdout.write;
+  const originalStderrWrite = process.stderr.write;
+  testServerConnection.onStdio(chunk => {
+    const text = chunk.text ?? Buffer.from(chunk.buffer!, 'base64');
+    if (chunk.type === 'stderr')
+      originalStderrWrite.call(process.stderr, text);
+    else
+      originalStdoutWrite.call(process.stdout, text);
+  });
+
+  await testServerConnection.initialize({ interceptStdio: true });
```

Comment:
Why do we want to intercept stdio, when we immediately write it back?

Reply:
good call, removed that.